### PR TITLE
Add ROIRandomCrop operator

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed to perform biased random crop in segmentation applications

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added a new operator ROIRandomCrop that takes a fixed cropping window shape and a region of interest and produces a cropping window anchor (start coordinates) so that the resulting cropping window contains as much of the ROI as possible.*
 - Affected modules and functionalities:
     *New op*
 - Key points relevant for the review:
     *Calculation of crop anchor.*
     *Test coverage? corner cases?*
 - Validation and testing:
     *Python tests*
 - Documentation (including examples):
     *Operator documentation*

**JIRA TASK**: *[DALI-1816]*

### CODE REVIEW

File: dali/operators/generic/roi_random_crop.cc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+#include <random>
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/pipeline/operator/arg_helper.h"
+
+namespace dali {
+
+DALI_SCHEMA(ROIRandomCrop)
+    .DocStr(R"code(Produces a fixed shape cropping window, randomly placed so that as much of the
+provided region of interest (ROI) is contained in it.
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:24]:
```suggestion
provided region of interest (ROI) is contained in it.

```

File: dali/operators/generic/roi_random_crop.cc [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    .AddOptionalArg<std::vector<int>>("roi_end",
+      R"code(ROI end coordinates.
+
+Note: Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("roi_shape",
+      R"code(ROI shape.
+
+Note: Using ``roi_shape`` is mutually exclusive with ``roi_end``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("in_shape",
+      R"code(Shape of the input data.
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:53]:
```suggestion
.. Note::

     Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
    .AddOptionalArg<std::vector<int>>("roi_shape",
      R"code(ROI shape.

.. Note::

    Using ``roi_shape`` is mutually exclusive with ``roi_end``.)code", nullptr, true)
    .AddOptionalArg<std::vector<int>>("in_shape",
      R"code(Shape of the input data.
```

File: dali/operators/generic/roi_random_crop.cc [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    .AddArg("roi_start",
+      R"code(ROI start coordinates.)code", DALI_INT_VEC, true)
+    .AddOptionalArg<std::vector<int>>("roi_end",
+      R"code(ROI end coordinates.
+
+Note: Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("roi_shape",
+      R"code(ROI shape.
+
+Note: Using ``roi_shape`` is mutually exclusive with ``roi_end``.)code", nullptr, true)
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:51]:
```suggestion
.. Note::

     Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
    .AddOptionalArg<std::vector<int>>("roi_shape",
      R"code(ROI shape.

.. Note::

    Using ``roi_shape`` is mutually exclusive with ``roi_end``.)code", nullptr, true)
```

File: dali/operators/generic/roi_random_crop.cc [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    .AddOptionalArg<std::vector<int>>("in_shape",
+      R"code(Shape of the input data.
+
+If provided, the cropping window start will be selected so that the cropping window is within the
+bounds of the input.
+
+Note: Providing ``in_shape`` is incompatible with feeding the input data directly as a positional input.
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:58]:
```suggestion
.. Note::

    Providing ``in_shape`` is incompatible with feeding the input data directly as a positional input.
```

File: dali/operators/generic/roi_random_crop.cc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    .AddArg("crop_shape",
+      R"code(Cropping window dimensions.)code", DALI_INT_VEC, true)
+    .AddArg("roi_start",
+      R"code(ROI start coordinates.)code", DALI_INT_VEC, true)
+    .AddOptionalArg<std::vector<int>>("roi_end",
+      R"code(ROI end coordinates.
+
+Note: Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("roi_shape",
+      R"code(ROI shape.
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:49]: So ROI shape doesn't have to be aligned with `in_shape`?

File: dali/operators/generic/roi_random_crop.cc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+ROIRandomCropCPU::ROIRandomCropCPU(const OpSpec &spec)
+    : Operator<CPUBackend>(spec),
+      rngs_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<int64_t>("max_batch_size")),
+      roi_start_("roi_start", spec),
+      roi_end_("roi_end", spec),
+      roi_shape_("roi_shape", spec),
+      crop_shape_("crop_shape", spec),
+      in_shape_arg_("in_shape", spec) {
+  if ((roi_end_.IsDefined() + roi_shape_.IsDefined()) != 1)
+    DALI_FAIL("Either ROI end or ROI shape should be defined, but not both");
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:93]:
```suggestion
  DALI_ENFORCE(roi_end_.IsDefined() + roi_shape_.IsDefined()) == 1,
                            "Either ROI end or ROI shape should be defined, but not both");
```

File: dali/operators/generic/roi_random_crop.cc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+bool ROIRandomCropCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<CPUBackend> &ws) {
+  int nsamples = spec_.HasTensorArgument("crop_shape") ?
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:98]: Maybe we should enforce that either `crop_shape` or input provided. Now if we won't provide anything it will crash (I guess).

File: dali/operators/generic/roi_random_crop.cc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                     ws.GetRequestedBatchSize(0);
+  crop_shape_.Acquire(spec_, ws, nsamples, true);
+  int ndim = crop_shape_[0].shape[0];
+
+  TensorShape<1> sh{ndim};
+  roi_start_.Acquire(spec_, ws, nsamples, sh);
+  if (roi_end_.IsDefined()) {
+    roi_end_.Acquire(spec_, ws, nsamples, sh);
+  } else {
+    assert(roi_shape_.IsDefined());
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:109]: Do you need that assert as you already have one check in the constructor?

File: dali/operators/generic/roi_random_crop.cc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    roi_shape_.Acquire(spec_, ws, nsamples, sh);
+  }
+
+  in_shape_.shapes.clear();
+  if (in_shape_arg_.IsDefined() || ws.NumInput() == 1) {
+    if (in_shape_arg_.IsDefined() && (ws.NumInput() == 1)) {
+      DALI_FAIL("``in_shape`` argument is incompatible with providing an input.")
+    }
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:117]:
```suggestion
   DALI_ENFORCE(in_shape_arg_.IsDefined() && (ws.NumInput() == 1) == false, "``in_shape`` argument is incompatible with providing an input.")
```

File: dali/operators/generic/roi_random_crop.cc [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+      auto sample_sh = in_shape_.tensor_shape_span(s);
+      for (int d = 0; d < ndim; d++) {
+        DALI_ENFORCE(sample_sh[d] >= crop_shape_[s].data[d],
+                     "Cropping shape can't be bigger than the input shape.");
+      }
+      if (roi_shape_.IsDefined()) {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 &&
+                       sample_sh[d] >= (roi_start_[s].data[d] + roi_shape_[s].data[d]),
+                       "ROI can't be out of bounds.");
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:142]: Can you print the offending shape and roi bounds?

File: dali/test/python/test_operator_roi_random_crop.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            elif use_in_shape_arg:
+                in_shape_arg = dali.fn.external_source(shape_gen_f, batch=False)
+                in_shape_out = in_shape_arg
+        crop_shape = fn.random.uniform(range=(crop_min_extent, crop_max_extent + 1), 
+                                       shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_shape = fn.random.uniform(range=(roi_min_extent, roi_max_extent + 1),
+                                      shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_start = fn.random.uniform(range=(roi_min_start, roi_max_start + 1),
+                                      shape=(ndim,), dtype=types.INT32, device='cpu')
+        crop_start = fn.roi_random_crop(*inputs, crop_shape=crop_shape,
```
Comment (JanuszL) [inline dali/test/python/test_operator_roi_random_crop.py:47]: I think you should test `roi_end` argument as well.

File: dali/test/python/test_operator_roi_random_crop.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+                                        roi_start=roi_start, roi_shape=roi_shape,
```
Comment (JanuszL) [inline dali/test/python/test_operator_roi_random_crop.py:48]: Can you add a negative cases when both `use_in_shape_arg` and `use_shape_like_in` are provided, as well as `roi_end` with `roi_shape`?

File: dali/test/python/test_operator_roi_random_crop.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                                        in_shape=in_shape_arg, device='cpu')
+    outputs = [roi_start, roi_shape, crop_start, crop_shape]
+    if use_in_shape_arg or use_shape_like_in:
+        outputs += [in_shape_out]
+    pipe.set_outputs(*outputs)
+    pipe.build()
+    for iter in range(niter):
```
Comment (JanuszL) [inline dali/test/python/test_operator_roi_random_crop.py:55]:
```suggestion
    for _ in range(niter):
```

File: dali/test/python/test_operator_roi_random_crop.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        outputs = pipe.run()
+        for idx in range(batch_size):
+            roi_start = outputs[0].at(idx).tolist()
```
Comment (JanuszL) [inline dali/test/python/test_operator_roi_random_crop.py:58]: Can you not use `at` as it is deprecated?

File: dali/test/python/test_operator_roi_random_crop.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            in_shape = None
+            if use_in_shape_arg or use_shape_like_in:
+                in_shape = outputs[4].at(idx).tolist()
+
+            roi_end = [roi_start[d] + roi_shape[d] for d in range(ndim)]
+            crop_end = [crop_start[d] + crop_shape[d] for d in range(ndim)]
+
+            for d in range(ndim):
+                if in_shape is not None:
+                    assert crop_start[d] >= 0
```
Comment (JanuszL) [inline dali/test/python/test_operator_roi_random_crop.py:72]: Shouldn't always be `crop_start[d] >= 0` no matter if `in_shape` is provided?

File: dali/operators/generic/roi_random_crop.cc [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+      for (int s = 0; s < nsamples; s++) {
+        auto sample_sh = in_shape_.tensor_shape_span(s);
+        for (int d = 0; d < ndim; d++) {
+          sample_sh[d] = in_shape_arg_[s].data[d];
+        }
+      }
+    } else {
+      auto &in = ws.template InputRef<CPUBackend>(0);
+      in_shape_ = in.shape();
+    }
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:130]: What happens if no input nor `in_shape_arg_` are provided?

File: dali/operators/generic/roi_random_crop.cc [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+      }
+      if (roi_shape_.IsDefined()) {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 &&
+                       sample_sh[d] >= (roi_start_[s].data[d] + roi_shape_[s].data[d]),
+                       "ROI can't be out of bounds.");
+        }
+      } else {
+        for (int d = 0; d < ndim; d++) {
+          DALI_ENFORCE(roi_start_[s].data[d] >= 0 && sample_sh[d] >= roi_end_[s].data[d],
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:146]: Do we need to bother with `roi_end_` here. Maybe it should be just `roi_start_` and `roi_shape_`, and values from `roi_end` should be used to infer `roi_shape_` if not provided directly.

File: dali/operators/generic/roi_random_crop.cc [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    DALI_FAIL("Either ROI end or ROI shape should be defined, but not both");
+}
+
+bool ROIRandomCropCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<CPUBackend> &ws) {
+  int nsamples = spec_.HasTensorArgument("crop_shape") ?
+                     ws.ArgumentInput("crop_shape").size() :
+                     ws.GetRequestedBatchSize(0);
+  crop_shape_.Acquire(spec_, ws, nsamples, true);
+  int ndim = crop_shape_[0].shape[0];
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:102]: Maybe you should check if `crop_shape_` is uniform?

File: dali/operators/generic/roi_random_crop.cc [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+    thread_pool.AddWork(
+      [&, ndim, sample_idx](int thread_id) {
+        int64_t* sample_sh = nullptr;
+        if (!in_shape_.empty())
+          sample_sh = in_shape_.tensor_shape_span(sample_idx).data();
+        for (int d = 0; d < ndim; d++) {
+          int64_t roi_extent = -1;
+          int64_t roi_start = roi_start_[sample_idx].data[d];
+          int64_t crop_extent = crop_shape_[sample_idx].data[d];
+          if (roi_end_.IsDefined()) {
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:178]: I think you can do this in setup.

File: dali/operators/generic/roi_random_crop.cc [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+            if (sample_sh)
+              range_end = std::min<int64_t>(sample_sh[d] - crop_extent, range_end);
+            auto dist = std::uniform_int_distribution<int64_t>(roi_start, range_end);
+            crop_start[sample_idx].data[d] = dist(rngs_[sample_idx]);
+          } else {  // roi_extent < crop_extent
+            int64_t range_start = roi_start + roi_extent - crop_extent;
+            if (sample_sh)
+              range_start = std::max<int64_t>(0, range_start);
+            auto dist = std::uniform_int_distribution<int64_t>(range_start, roi_start);
+            crop_start[sample_idx].data[d] = dist(rngs_[sample_idx]);
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:197]:
```suggestion
            int64_t range = crop_extent - roi_extent;
            if (sample_sh)
              range = std::max<int64_t>(0, roi_start - range);
            auto dist = std::uniform_int_distribution<int64_t>(0, range);
            crop_start[sample_idx].data[d] = roi_start - dist(rngs_[sample_idx]);
```
Just a suggestion, it may better reflect what actually happens.

File: dali/operators/generic/roi_random_crop.cc [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+          } else {
+            roi_extent = roi_shape_[sample_idx].data[d];
+          }
+
+          if (roi_extent == crop_extent) {
+            crop_start[sample_idx].data[d] = roi_start;
+          } else if (roi_extent > crop_extent) {
+            int64_t range_end = roi_start + roi_extent - crop_extent;
+            if (sample_sh)
+              range_end = std::min<int64_t>(sample_sh[d] - crop_extent, range_end);
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:189]: Do we need that. I think Setup already checks if ROI is inside shape bounds.

File: dali/operators/generic/roi_random_crop.cc [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+          int64_t crop_extent = crop_shape_[sample_idx].data[d];
+          if (roi_end_.IsDefined()) {
+            roi_extent = roi_end_[sample_idx].data[d] - roi_start;
+          } else {
+            roi_extent = roi_shape_[sample_idx].data[d];
+          }
+
+          if (roi_extent == crop_extent) {
+            crop_start[sample_idx].data[d] = roi_start;
+          } else if (roi_extent > crop_extent) {
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:186]: How about:
```suggestion
          if (roi_extent == crop_extent) {
            crop_start[sample_idx].data[d] = roi_start;
          } else {
            int64_t range = roi_extent - crop_extent;
            if (sample_sh) {
              range = std::max<int64_t>(range, -roi_start); // make sure that if roi_extent < crop_extent we won't be < 0
            }
            int64_t range_start = std::min<int64_t>(roi_start, roi_start + range);
            int64_t range_end = std::max<int64_t>(roi_start, roi_start + range)
            auto dist = std::uniform_int_distribution<int64_t>(range_start, range_end);
            crop_start[sample_idx].data[d] = dist(rngs_[sample_idx]);
         }
```

File: dali/operators/generic/roi_random_crop.cc (hunk 21 above)
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:189]: I think you check if ROI is inside shape already in setup. So this is not needed.

File: dali/operators/generic/roi_random_crop.cc (hunk 22 above)
Comment (JanuszL) [reply dali/operators/generic/roi_random_crop.cc:186]: Maybe you can get rid of if at all.

File: dali/operators/generic/roi_random_crop.cc [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+#include "dali/pipeline/operator/arg_helper.h"
+
+namespace dali {
+
+DALI_SCHEMA(ROIRandomCrop)
+    .DocStr(R"code(Produces a fixed shape cropping window, randomly placed so that as much of the
+provided region of interest (ROI) is contained in it.
+
+If the ROI is bigger than the cropping window, the cropping window will be a subwindow of the ROI.
+If the ROI is smaller than the cropping window, the whole ROI should be contained in the cropping window.
```
Comment (mzient) [inline dali/operators/generic/roi_random_crop.cc:27]:
```suggestion
If the ROI is smaller than the cropping window, the whole ROI shall be contained in the cropping window.
```

File: dali/operators/generic/roi_random_crop.cc [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+
+If an input shape (``in_shape``) is given, the resulting cropping window is selected to be within the
+bounds of that input shape. Alternatively, the input data subject to cropping can be fed directly to
```
Comment (mzient) [inline dali/operators/generic/roi_random_crop.cc:30]:
```suggestion
bounds of that input shape. Alternatively, the input data subject to cropping can be passed to the operator, in
which case the cropping window will be constrained to the shape of the data.
```
`fed` suggests that the data itself is consumed - and I think we should make it very clear that the operator doesn't touch the data.

File: dali/operators/generic/roi_random_crop.cc (hunk 18 above)
Comment (mzient) [reply dali/operators/generic/roi_random_crop.cc:109]: It doesn't have to - each sample can be different.

File: dali/operators/generic/roi_random_crop.cc (hunk 8 above)
Comment (mzient) [reply dali/operators/generic/roi_random_crop.cc:116]: Isn't it one of the reasons to have asserts - to check for inconsistent internal state?

File: dali/operators/generic/roi_random_crop.cc (hunk 17 above)
Comment (mzient) [reply dali/operators/generic/roi_random_crop.cc:162]: It's easier to check `end` than `start+shape`?

File: dali/operators/generic/roi_random_crop.cc [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+        DALI_ENFORCE(sample_sh[d] >= 0,
+                     make_string("Input shape can't be negative. Got ",
+                                 sample_sh[d], " for d=", d));
+        DALI_ENFORCE(crop_shape_[s].data[d] >= 0,
+                     make_string("Crop shape can't be negative. Got ", crop_shape_[s].data[d],
+                                 " for d=", d));
+        DALI_ENFORCE(sample_sh[d] >= crop_shape_[s].data[d],
+                     make_string("Cropping shape can't be bigger than the input shape. "
+                                 "Got: crop_shape[", crop_shape_[s].data[d], "] and sample_shape[",
+                                 sample_sh[d], "] for d=", d));
```
Comment (mzient) [inline dali/operators/generic/roi_random_crop.cc:150]: I wonder what we should do in this case...

File: dali/operators/generic/roi_random_crop.cc [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+  auto &out_crop_start = ws.template OutputRef<CPUBackend>(0);
+  auto crop_start = view<int64_t, 1>(out_crop_start);
+
+  int nsamples = crop_start.shape.size();
+  int ndim = crop_start[0].shape[0];
+
+  auto& thread_pool = ws.GetThreadPool();
+
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    thread_pool.AddWork(
```
Comment (mzient) [inline dali/operators/generic/roi_random_crop.cc:187]: I wonder if asynchronous execution is justified here - after all this operator performs `ndim` loops per sample - not much compared to the overhead of scheduling work to thread pool.

File: dali/operators/generic/roi_random_crop.cc (hunk 25 above)
Comment (mzient) [reply dali/operators/generic/roi_random_crop.cc:150]: Maybe we should allow (with an extra parameter) to clip the cropping window if it's larger than input? Or, alternatively, keep out-of-bounds window (pad)? Just thinking aloud.

File: dali/operators/generic/roi_random_crop.cc (hunk 8 above)
Comment (jantonguirao) [reply dali/operators/generic/roi_random_crop.cc:116]: I put it there mostly for documentation, but also to double-check that the logic is sane

File: dali/operators/generic/roi_random_crop.cc (hunk 25 above)
Comment (jantonguirao) [reply dali/operators/generic/roi_random_crop.cc:150]: I don't think we should complicate this even more unless we have a specific use in mind.

File: dali/operators/generic/roi_random_crop.cc (hunk 18 above)
Comment (jantonguirao) [reply dali/operators/generic/roi_random_crop.cc:109]: This is the shape of a shape we are talking about. Uniformity in crop_shape means same number of dimensions

File: dali/operators/generic/roi_random_crop.cc (hunk 26 above)
Comment (jantonguirao) [reply dali/operators/generic/roi_random_crop.cc:187]: I can get rid of it

File: dali/operators/generic/roi_random_crop.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/operators/generic/roi_random_crop.cc:24]: done

File: dali/test/python/test_operator_roi_random_crop.py [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+                        assert crop_end[d] >= roi_end[d]
+                    else:
+                        assert crop_start[d] >= roi_start[d]
+                        assert crop_end[d] <= roi_end[d]
+            for idx in range(4, 6):
+                check_crop_start(np.array(outputs[idx][s]).tolist(), roi_start, roi_shape, crop_shape)
+            for idx in range(6, 10):
+                check_crop_start(np.array(outputs[idx][s]).tolist(), roi_start, roi_shape, crop_shape, in_shape)
+
+def test_random_mask_pixel():
```
Comment (mzient) [inline dali/test/python/test_operator_roi_random_crop.py:92]:
```suggestion
def test_roi_random_cropl():
```
?

File: dali/test/python/test_operator_roi_random_crop.py [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+        shape_like_in = dali.fn.external_source(lambda: np.zeros(shape_gen_f()),
+                                                device='cpu', batch=False)
+        in_shape = dali.fn.shapes(shape_like_in, dtype=types.INT32)
+
+        crop_shape = fn.random.uniform(range=(crop_min_extent, crop_max_extent + 1), 
+                                       shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_shape = fn.random.uniform(range=(roi_min_extent, roi_max_extent + 1),
+                                      shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_start = fn.random.uniform(range=(roi_min_start, roi_max_start + 1),
+                                      shape=(ndim,), dtype=types.INT32, device='cpu')
```
Comment (mzient) [inline dali/test/python/test_operator_roi_random_crop.py:37]: I see that scalar arguments are not tested. Perhaps you could do sth like:
```Python
def make_arg(lo, hi):
  return [lo] * ndim if lo == hi else fn.random.uniform(range=(lo, hi+1), shape=[ndim], dtype=types.INT32)
```

File: dali/test/python/test_operator_roi_random_crop.py [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+                          roi_min_extent = 20, roi_max_extent = 50,
+                          crop_min_extent = 20, crop_max_extent = 50,
+                          in_shape_min = 400, in_shape_max = 500,
+                          niter=3):
+    pipe = dali.pipeline.Pipeline(batch_size=batch_size, num_threads=4, device_id=0, seed=1234)
+    with pipe:
+        assert in_shape_min < in_shape_max
+        shape_gen_f = lambda: random_shape(in_shape_min, in_shape_max, ndim)
+        shape_like_in = dali.fn.external_source(lambda: np.zeros(shape_gen_f()),
+                                                device='cpu', batch=False)
```
Comment (mzient) [inline dali/test/python/test_operator_roi_random_crop.py:29]: Try variable batch size?

File: dali/test/python/test_operator_roi_random_crop.py (hunk 27 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_roi_random_crop.py:92]: yes, copy-paste :)

File: dali/operators/generic/roi_random_crop.cc (hunk 16 above)
Comment (JanuszL) [reply dali/operators/generic/roi_random_crop.cc:136]: Now I see. I missed that L121 condition covers the L138 loop as well.

File: dali/operators/generic/roi_random_crop.cc [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+void ROIRandomCropCPU::RunImpl(workspace_t<CPUBackend> &ws) {
+  auto &out_crop_start = ws.template OutputRef<CPUBackend>(0);
+  auto crop_start = view<int64_t, 1>(out_crop_start);
+
+  int nsamples = crop_start.shape.size();
+  int ndim = crop_start[0].shape[0];
+
+  auto& thread_pool = ws.GetThreadPool();
```
Comment (JanuszL) [inline dali/operators/generic/roi_random_crop.cc:184]: Remove

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/test_operator_roi_random_crop.py [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+            else  fn.random.uniform(range=(crop_min_extent, crop_max_extent + 1), 
+                                    shape=(ndim,), dtype=types.INT32, device='cpu')
+
+        roi_shape = np.array([(roi_min_extent + roi_max_extent) // 2] * ndim, dtype=np.int32) if random.choice([True, False]) \
+            else fn.random.uniform(range=(roi_min_extent, roi_max_extent + 1),
+                                    shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_start = np.array([(roi_min_start + roi_max_start) // 2] * ndim, dtype=np.int32) if random.choice([True, False]) \
+            else fn.random.uniform(range=(roi_min_start, roi_max_start + 1),
+                                    shape=(ndim,), dtype=types.INT32, device='cpu')
+        roi_end = roi_start + roi_shape
```
Comment (mzient) [inline dali/test/python/test_operator_roi_random_crop.py:50]: If this works, then it's by chance - if `roi_start` is a `numpy.ndarray` and `roi_shape` is a `DataNode`, this wouldn't work.

File: dali/test/python/test_operator_roi_random_crop.py [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+                          niter=3):
+    pipe = dali.pipeline.Pipeline(batch_size=max_batch_size, num_threads=4, device_id=0, seed=1234)
+    with pipe:
+        assert in_shape_min < in_shape_max
+        shape_gen_fn = lambda: random_shape(in_shape_min, in_shape_max, ndim)
+        data_gen_f = lambda: batch_gen(max_batch_size, shape_gen_fn)
+        shape_like_in = dali.fn.external_source(data_gen_f, device='cpu')
+        in_shape = dali.fn.shapes(shape_like_in, dtype=types.INT32)
+
+        crop_shape =  [(crop_min_extent + crop_max_extent) // 2] * ndim if random.choice([True, False]) \
```
Comment (mzient) [inline dali/test/python/test_operator_roi_random_crop.py:40]: Numpy arrays are passed as argument inputs, so this still does not test non-tensor arguments. It must be a plain list.

Comment (jantonguirao) [reply dali/test/python/test_operator_roi_random_crop.py:40]: You are right. I fixed that now

**Review by mzient — APPROVED** [review]
